# Convolutional Layers in Keras

Convolution with 3x3 window and stride 1:

<img src='conv-visualization/images/full-padding-no-strides-transposed.gif'>

Image source: http://iamaaditya.github.io/2016/03/one-by-one-convolution/

To create a convolutional layer in Keras, we must first import the necessary module:



In [5]:
from keras.layers import Conv2D

Then, we can create a convolutional layer by using the following format:

In [ ]:
Conv2D(filters, kernel_size, strides, padding, activation='relu', input_shape)

## Arguments

We must pass the following arguments:

- `filters` - The number of filters.
- `kernel_size` - Number specifying both the height and width of the (square) convolution window.

There are some additional, optional arguments that we might like to tune:

- `strides` - The stride of the convolution. If we do not specify anything, strides is set to 1.
- `padding` - One of `valid` or `same`. If we do not specify anything, padding is set to `valid`.
- `activation` - Typically `relu`. If we do not specify anything, no activation is applied. We are strongly encouraged to add a ReLU activation function to every convolutional layer in your networks.

NOTE: It is possible to represent `both kernel_size` and `strides` as either a number or a tuple.



When using our convolutional layer as the first layer (appearing after the input layer) in a model, we must provide an additional `input_shape` argument:

- `input_shape` - Tuple specifying the height, width, and depth (in that order) of the input.

NOTE: Do not include the `input_shape` argument if the convolutional layer is not the first layer in your network.

There are many other tunable arguments that we can set to change the behavior of our convolutional layers. To read more about these, we recommend perusing the official [documentation](https://keras.io/layers/convolutional/).

### Example 1

Say we are constructing a CNN, and out input layer accepts grayscale images that are 200 by 200 pixels (corresponding to a 3D array with height 200, width 200, and depth 1). Then, say we would like the next layer to be a convolutional layer with 16 filters, each with a width and height of 2. When performing the convolution, we  would like the filter to jump two pixels at a time. We also do not want the filter to extend outside of the image boundaries; in other words, we do not want to pad the image with zeros. Then, to construct this convolutional layer, we would use the following line of code:

In [ ]:
Conv2D(filters=16, kernel_size=2, strides=2, activation='relu', input_shape=(200, 200, 1))

### Example 2

Say we would like the next layer in my CNN to be a convolutional layer that takes the layer constructed in Example 1 as input. Say we would like our new layer to have 32 filters, each with a height and width of 3. When performing the convolution, we would like the filter to jump 1 pixel at a time. We want the convolutional layer to see all regions of the previous layer, and so we do not mind if the filter hangs over the edge of the previous layer when it is performing the convolution. Then, to construct this convolutional layer, we would use the following line of code:

In [ ]:
Conv2D(filters=32, kernel_size=3, padding='same', activation='relu')

### Example 3

If we look up code online, it is also common to see convolutional layers in Keras in this format:

In [ ]:
Conv2D(64, (2,2), activation='relu')

In this case, there are 64 filters, each with a size of 2x2, and the layer has a ReLU activation function. The other arguments in the layer use the default values, so the convolution uses a stride of 1, and the padding has been set to `valid`.

## Dimensionality

Convolution with 3x3 window and stride 1:

<img src='conv-visualization/images/convolution-schematic.gif'>

Image source: http://deeplearning.stanford.edu/wiki/index.php/Feature_extraction_using_convolution

Just as with neural networks, we create a CNN in Keras by first creating a Sequential model.

We add layers to the network by using the `.add()` method. Consider the following CNN:

In [7]:
from keras.models import Sequential
from keras.layers import Conv2D

model = Sequential()
model.add(Conv2D(filters=16, kernel_size=2, strides=2, padding='valid', 
    activation='relu', input_shape=(200, 200, 1)))
model.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_1 (Conv2D)            (None, 100, 100, 16)      80        
Total params: 80
Trainable params: 80
Non-trainable params: 0
_________________________________________________________________


Do the dimensions of the convolutional layer line up with our expectations?

Feel free to change the values assigned to the arguments (`filters`, `kernel_size`, etc).

Take note of how the number of parameters in the convolutional layer changes. This corresponds to the value under `Param #` in the printed output.

Also notice how the shape of the convolutional layer changes. This corresponds to the value under `Output Shape` in the printed output. It follows the format: (batch size, height, width, depth).

### Formula: Number of Parameters in a Convolutional Layer

The number of parameters in a convolutional layer depends on the supplied values of `filters`, `kernel_size`, and `input_shape`. Let us define a few variables:

- `K` - the number of filters in the convolutional layer
- `F` - the height and width of the convolutional filters
- `D_in` - the depth of the previous layer

Notice that `K` = `filters`, and `F` = `kernel_size`. Likewise, `D_in` is the last value in the `input_shape` tuple.

Since there are `F*F*D_in` weights per filter, and the convolutional layer is composed of `K` filters, the total number of weights in the convolutional layer is `K*F*F*D_in`. Since there is one bias term per filter, the convolutional layer has `K` biases. Thus, the number of parameters in the convolutional layer is given by `K*F*F*D_in + K`.

### Formula: Shape of a Convolutional Layer

The shape of a convolutional layer depends on the supplied values of `kernel_size`, `input_shape`, `padding` and `stride`. Let us define a few variables:

- `K` - the number of filters in the convolutional layer
- `F` - the height and width of the convolutional filters
- `S` - the stride of the convolution
- `H_in` - the height of the previous layer
- `W_in` - the width of the previous layer

Notice that `K` = `filters`, `F` = `kernel_size`, and `S` = `stride`. Likewise, `H_in` and `W_in` are the first and second value of the `input_shape tuple`, respectively.

The depth of the convolutional layer will always equal the number of filters `K`.

If `padding` =  `same`, then the spatial dimensions of the convolutional layer are the following:

- `height` = `math.ceil(float(H_in) / float(S))`
- `width` = `math.ceil(float(W_in) / float(S))`

If `padding` = `valid`, then the spatial dimensions of the convolutional layer are the following:

- `height` = `math.ceil(float(H_in - F + 1) / float(S))`
- `width` = `math.ceil(float(W_in - F + 1) / float(S))`

*Note*: Information about the `ceil()` method can be found in the [documentation](https://docs.python.org/3/library/math.html) and some python [tutorials](https://www.tutorialspoint.com/python/number_ceil.htm).

### Quiz

In [8]:
model = Sequential()
model.add(Conv2D(filters=32, kernel_size=3, strides=2, padding='same', 
    activation='relu', input_shape=(128, 128, 3)))
model.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_2 (Conv2D)            (None, 64, 64, 32)        896       
Total params: 896
Trainable params: 896
Non-trainable params: 0
_________________________________________________________________


1. How many parameters does this convolutional layer have?

    The number of parameters is (32 x 3 x 3 x 3) + 32 = 896, just as shown under `Param #`.
    
2. What is the depth of the convolutional layer?

    The depth of a convolutional layer is always equal to the number of filters, which is 32.
    
3. What is the width of the convolutional layer?

    The width of the convolutional layer is 64, as can be seen from the second element of the `Output Shape` object.